Initial KEP-2021 draft

johanneswuerbach opened the pull request:
Enhancement issue: https://github.com/kubernetes/enhancements/issues/2021

kikisdeliveryservice reviewed (commented):
Hi @johanneswuerbach 

Enhancements Lead here, some notes on the the current PR: 

The dir structure should be: keps/sig-autoscaling/2021-scale-from-zero/  (Note we put the issue number in the title now)

The current file 20200926-scale-from-zero.md should be renamed README.md and a kep.yaml should also be added.

More detail can be found here: https://github.com/kubernetes/enhancements/tree/master/keps/NNNN-kep-template

Best,
Kirsten

johanneswuerbach commented:
@kikisdeliveryservice thank you, folder structure fixed.

johanneswuerbach commented:
/assign @gjtempleton

I hoped to get the current state document first, before we start talking about how we could move towards beta. Does that make sense?

gjtempleton commented:
Hey @johanneswuerbach 

I think that's a good plan for now.

jeffreybrowning commented:
@johanneswuerbach @gjtempleton how's this going so far? Can I be of any help?

johanneswuerbach commented:
@jeffreybrowning yes. I assumed we could merge this KEP already as is to document the current state and then iterate on it towards beta. I'll try to present it at the next sig-autoscaling meeting to get some input and discuss next steps, but if you have any input I'm happy to incorporate it already.

jeffreybrowning commented:
@johanneswuerbach missed autoscaling meeting today -- do you have clarity on next steps?

johanneswuerbach commented:
Me too, I assumed those are bi-weekly or are the meetings on-demand?

jeffreybrowning commented:
In all honesty, it would have been my first one -- the work you started on this feature for beta has encouraged me to get involved and help you push this through. 

It will really help using HPA + async job queue to scale down to 0 workers when not processing tasks.

gjtempleton commented:
Hey, the meetings are held weekly at 14:00 UTC, you can see more info including a link to the agenda if you want to raise it [here](https://github.com/kubernetes/community/tree/master/sig-autoscaling). 

I've raised this at a previous meeting, so the community's already aware this work has started, but it would be good to have some more in-depth discussion.

johanneswuerbach commented:
Thanks I got confused by the mention of biweekly here https://github.com/kubernetes/community/tree/master/sig-autoscaling#meetings and assumed the calendar invite is wrong. Will ask upfront next time :-)

jeffreybrowning commented:
Pinging here. How's this enhancement coming along? What are next steps?

jeffreybrowning commented:
Pinging back before the holidays hit. 

What are next steps?

johanneswuerbach commented:
Sorry for leaving this lingering @jeffreybrowning, it wasn't forgotten just work happened :-(

@gjtempleton during the SIG meeting you said merging the KEP only documenting the alpha state would be okay. I think that is done, or do you see anything missing?

For the beta stage the following needs to happen:

* Figure out a way to deal with k8s upgrade/downgrade when `minReplicas: 0`, currently this was worked around by the feature gate, but could look like https://github.com/kubernetes/kubernetes/pull/74526#issuecomment-497549411
* Resolve the issue that going from `minReplicas` `0` ~> `1` seems to not trigger a scale up and the scaled resource stays at 0 replicas.
* Figure out how to deal with the fact that HPA can no longer be disabled, which was previously done by setting `replicas: 0` on the scaled resource.

jeffreybrowning commented:
No worries @johanneswuerbach, and didn't want to rush you. Just think your work already is important and want to see it to Beta for this feature!

Asking some questions to spin myself up:

1. Is HPA with `minReplicas: 0` leading to disabling the HPA an intended feature? Is there known methodology behind wanting a way to disable an HPA rather than just removing it? It seems to me that not many k8s resources have intentional disabled states, and instead require deleting? 
2. `Resolve the issue that going from minReplicas 0 ~> 1 seems to not trigger a scale up and the scaled resource stays at 0 replicas.` Is there an issue for this already? Just to be clear, does this mean that the HPA increasing its recommended replica count does not cause the scaled resource to scale? 

The recommendations on upgrate to Beta on how to handle downgrades makes sense: since downgrades would require the feature-gate to be set manually, than moving down to version with the feature-gate requirement will not be a surprise.

MattJeanes commented:
Can we remove the stale state on this? I think this is a very important and useful feature for kubernetes

gjtempleton commented:
/remove-lifecycle stale

johanneswuerbach commented:
Sorry for the lack of updates, I pushed a refined proposal around handling disabling and updating from `0` to `1`.

Let me know what you think.

elmiko commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,660 @@
+## Summary
+
+[Horizontal Pod Autoscaler][] (HPA) automatically scales the number of pods in any resource which supports the `scale` subresource based on observed CPU utilization
+(or, with custom metrics support, on some other application-provided metrics) from one to many replicas. This proposal adds support for scaling from zero to many replicas
```
is this just about scaling "from" zero? it seems like this should mention both scaling "from" zero as well as "to" zero.

elmiko commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,660 @@
+implementation. The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
```
i think there should be a description of what the proposal is here. although i see the design details about adding "suspend" in the later sections, it would be nice to have a clear description of what is proposed and how it will utilized by users.

elmiko reviewed (commented):
i understand what the intended effect of this enhancement is, and the user story helps, but i could use a little more description about how this will work.

in specific i'm a little confused about the interaction of `minReplicas: 0` and the new `suspend` field. i think a little more description in the proposal section would help.

also, i think it would be nice to start trimming template comments out of this doc.

jeffreybrowning commented:
Going to risk bumping the thread here since its been a few weeks. @elmiko

elmiko commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,681 @@
+
+Additionally integration tests should be added for enable scale to zero by, setting
+ `enableScaleToZero: true`, setting `enableScaleToZero: true` and waiting for `replicas` to become `0` and another test for increasing `minReplicas: 1`, setting `enableScaleToZero: false` and observing that `replicas` became `1` again.
+
```
something that occurs to me while re-reading this section, what happens if the user changes `enableScaleToZero` while either `replicas` or `minreplicas` has been set to `0`. would there be any need to test these type of failure states?

johanneswuerbach replied:
I think those cases should be covered by controller unit tests. Do you think it is worth explicitly calling them out here?

josephburnett replied:
We should call them out here.  This test plan is for the whole feature, so we should list tests which should be implemented at every level (api, admission, controller, e2e).

josephburnett replied:
After merging this KEP we should come back the fill out the test plan a bit. Just to make sure we haven't missed anything.

elmiko reviewed (commented):
this looks good to me, i have one question that i didn't think of before, but otherwise this reads well to me.

MattJeanes commented:
/remove-lifecycle stale

Not sure if I have permission to do that but I'll give it a go.

This feature I believe remains a great addition to Kubernetes and while this can be achieved with something like KEDA, I think something like this should be a part of the core Kubernetes experience.

josephburnett commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,681 @@
+Additionally integration tests should be added for enable scale to zero by, setting
+ `enableScaleToZero: true`, setting `enableScaleToZero: true` and waiting for `replicas` to become `0` and another test for increasing `minReplicas: 1`, setting `enableScaleToZero: false` and observing that `replicas` became `1` again.
+
+### Graduation Criteria
```
We should have some graduation criteria.  How do we know when this feature can be promoted to "stable"?

johanneswuerbach replied:
I'm not really sure, except allowing for more feedback. Judging from the 👍  on this PR and the related issue there are folks waiting for this.

josephburnett replied:
Aside from verifying non-flakey E2E test coverage of the feature and anecdotal report from early production usage, I don't have any other ideas. So let's land this PR to achieve Beta. And we can always add GA criteria in another PR.

josephburnett commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,681 @@
+
+The production readiness review questionnaire must be completed for features in
+v1.19 or later, but is non-blocking at this time. That is, approval is not
+required in order to be in the release.
```
We'll have to do a production readiness review since "beta" features are usually enabled by default.

johanneswuerbach replied:
I would appreciate pointers here. Is this about filling out a questionnaire or more about getting a review by a specific person?

noahlz commented:
Where does this feature stand and how can the community help move it along? Currently my team is considering using https://github.com/dailymotion-oss/osiris as work-around, but given it's "experimental" we'd prefer to use this feature, should it be released.

JCMais commented:
@johanneswuerbach anything we can do to help you with this proposal?

johanneswuerbach commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,681 @@
+The `EnableScaleToZero` controls whether the `MinReplicas` can set to `>=0` (`true`, new behavior) or `>=1` (`false`, current behavior). The default will be `false` to preserve the current behavior.
+
+If `EnableScaleToZero` has been enabled, it can only be disabled when the scaled resource has at least one replica
+running and `MinReplicas` is `>=1`.
```
Unless we differentiate between `EnableScaleToZero=false` and `EnableScaleToZero=nil`, the controller can't determine in this case whether this is the backwards compatible "pause" case or a valid transition, because if the HPA wouldn't be updated, `minReplicas: 1`, `replicas: 0` means HPA is stopped.

josephburnett replied:
I see the dilemma.  We can unset enable scale to zero and set min replicas = 1, but if the scale target ref is still at zero, it's not going to scale up again.

So we need a tri-state: /1/ old behavior (replicas=0 means disabled) /2/ scale to zero enabled and /3/ scale to zero disabled (but replicas=0 will immediately scale to 1).  Having different behavior for nil versus false sounds confusing.

There is a new KEP which adds `Spec.UpdatePolicy.UpdateMode` of which `Off` and `Auto` are dry-run and normal modes respectively: https://github.com/kubernetes/kubernetes/pull/107998.  Dry run mode is essentially the old behavior when replicas = 0.  We could make the default `UpdateMode` something like "ZeroMeansOff" instead of `Auto` so we can express all three states.

```golang
const (
	// UpdateModeZeroMeanOff (default) means the autoscaler never creates/deletes pods
	// when the scale target ref returns replicas = 0.
	UpdateModeZeroMeansOff = "ZeroMeansOff"
	// UpdateModeOff means that autoscaler never creates/deletes pods.
	// The HorizontalPodAutoscalerStatus.DesiredReplicas field is still populated.
	// This can be used for a "dry run"
	UpdateModeOff UpdateMode = "Off"
	// UpdateModeAuto means that autoscaler creates/deletes pods based on the scaling decision
	// including scaling to zero replicas.
	UpdateModeAuto UpdateMode = "Auto"
)
```

@liggitt @mwielgus what do you think about this API change, combining scale-to-zero and dry-run?

liggitt replied:
Commented above, but removing the implicit pause on `replicas: 0` only for HPAs with `minReplicas: 0` makes more sense to me than a tri-state here.

josephburnett replied:
Okay. Let's keep the behavior as described then. You can resolve this comment.

johanneswuerbach commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,681 @@
+the various `minReplicas`, `replicas` and `enableScaleToZero` should be achievable with unit tests.
+
+Additionally integration tests should be added for enable scale to zero by, setting
+ `enableScaleToZero: true`, setting `enableScaleToZero: true` and waiting for `replicas` to become `0` and another test for increasing `minReplicas: 1`, setting `enableScaleToZero: false` and observing that `replicas` became `1` again.
```
Fixed.

johanneswuerbach commented:
/assign @josephburnett

c3ivodujmovic commented:
Just want to add a note of support for completing this as soon as possible. 
I think you are saying feature is beta in 1.25, and many in industry wish it was sooner. Last relevant code fix I could find was in 1.21 . I assume we cannot pronounce this beta in 1.23?

johanneswuerbach commented:
Sadly not, the existing design has some issues, which are planned to be addressed in the beta iteration. Afaik 1.24 is already closed for new features, so 1.25 seems to be the earliest target.

Where you can help? First this KEP needs a review to see whether there are any additional blockers otherwise I guess it would also be good to start with the implementation, if you have time.

liggitt commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,833 @@
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+From an UX perspective the two stage opt-out / opt-in from scale to zero might feel a bit tedious, but the only other available option seems to be deprecating the implicit HPA pause on `replicas: 0`. While this might provide an improved
+UX, it would require a full deprecation cycle (12 months) before graduating this feature from alpha to beta.
```
Removing the HPA pause on `replicas: 0` only for HPAs with `minReplicas: 0` makes sense to me. That's an opt-in on the HPA itself, would be documented as part of `scale-to-zero`, no existing HPAs (other than ones using the alpha scale-to-zero feature) would change behavior, and no 12-month deprecation is required.

raywainman replied:
@adrianmoisey @johanneswuerbach I really like this approach suggested by Jordan and it wouldn't require an additional enablement flag, simply setting `min_replicas = 0` is sufficient to "opt-into" this new behavior

cc @jm-franc

adrianmoisey replied:
I'm actually a user of the "scale replicas to 0 to disable the HPA" behaviour. I'm unsure how many other users there are, but if we remove the `EnableScaleToZero` opt-in, this means that a user needs to pick between "scale to zero" and "being able to disable the HPA".

We normally use the "scale replicas to 0 to disable the HPA" when we're fire-fighting an incident and want to turn off a bunch of background workers. So it's something we normally want to do quickly.

I do like @jm-franc's idea of a "disable" field, that might be a more obvious field than `EnableScaleToZero`.

"scale replicas to 0 to disable the HPA" has been around since 2016: https://github.com/kubernetes/kubernetes/pull/29212

raywainman replied:
+1 - I like the idea of HPA having a disable field rather than needing to change the replicas on the deployment. And this disable field would apply to every type of HPA.

What if you wanted to disable an HPA but NOT scale your workload down? Today you need to delete the HPA.

If we go with this suggestion then I believe we can proceed with this KEP without having to add all the complexity around the enablement flag.

adrianmoisey replied:
I can also see us splitting off the "Disabled" flag to its own KEP. Thoughts on that? I'm happy to run with that if needed.

omerap12 replied:
Yes. I like the idea of a "disable" field as well.

adrianmoisey replied:
(@liggitt I'm curious what your thoughts are regarding this)

johanneswuerbach replied:
Great idea, updated to use the condition instead of a new property.

zroubalik commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,833 @@
+-->
+
+## Alternatives
+
```
Might be worth noting that this behaviour can be achieve by KEDA.

noahlz replied:
You mean like https://keda.sh/docs/2.7/scalers/artemis/ ?

We use ArtemisMQ queue size as our metric and this is exactly our use case. We have several full-node pods (for large ETL jobs) that are idle ~20 hours / day.

gjtempleton replied:
Agreed, we should add this.

johanneswuerbach replied:
Added

MattJeanes commented:
/remove-lifecycle stale

This is still a valuable addition to Kubernetes I feel

josephburnett commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,833 @@
+
+#### Beta -> GA Graduation
+
+- Allowing time for feedback
```
Perhaps you could also mention E2E tests are passing without flakiness.

josephburnett commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,833 @@
+This section must be completed when targeting beta to a release.
+-->
+
+_This section must be completed when targeting beta graduation to a release._
```
Perhaps you could just mention there is no requirement for rollout and upgrade. And that rollback may require the operator to increase replicas from 0 manually (if that is correct).

josephburnett commented:
@johanneswuerbach I think we should merge this and proceed with Beta. I apologize for the delay in review. :(

One last request: please squash the commits into a single commit before we merge.

johanneswuerbach commented:
Thanks @josephburnett, I've squashed the commits and hopefully addressed your remaining comments.

josephburnett commented:
/lgtm
/approve

josephburnett commented:
/assign @deads2k
@gjtempleton will you approve this?

josephburnett commented:
@johanneswuerbach looks like you need to run `hack/update-toc.sh`.

johanneswuerbach commented:
> looks like you need to run `hack/update-toc.sh`.

done ✅

josephburnett commented:
/lgtm

kikisdeliveryservice commented on keps/sig-autoscaling/2021-scale-from-zero/kep.yaml:
```diff
@@ -0,0 +1,42 @@
+  - johanneswuerbach
+owning-sig: sig-autoscaling
+participating-sigs:
+status: provisional
```
if this is beta this shouldn't be provisional.

kikisdeliveryservice commented on keps/sig-autoscaling/2021-scale-from-zero/kep.yaml:
```diff
@@ -0,0 +1,42 @@
+# The most recent milestone for which work toward delivery of this KEP has been
+# done. This can be the current (upcoming) milestone, if it is being actively
+# worked on.
+latest-milestone: "v1.16"
```
this needs to be updated

kikisdeliveryservice commented on keps/sig-autoscaling/2021-scale-from-zero/kep.yaml:
```diff
@@ -0,0 +1,42 @@
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v1.16"
+  beta: "v1.25"
```
please update

kikisdeliveryservice commented on keps/prod-readiness/sig-autoscaling/2021.yaml:
```diff
@@ -0,0 +1,3 @@
+kep-number: 2021
+stable:
```
same here

johanneswuerbach commented:
@kikisdeliveryservice thanks for the review, comments addressed :-)

josephburnett commented:
/lgtm

mwielgus reviewed (approved):
/lgtm
/approve

josephburnett commented:
/assign @deads2k

josephburnett commented:
Ping @deads2k. We need your approval to merge this KEP.

pbetkier commented:
I believe @wojtek-t could also approve for us instead?

pbetkier commented:
/assign @johnbelamaric please approve for prod-readiness

pbetkier commented:
/assign @johnbelamaric

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,839 @@
+  Any change of default behavior may be surprising to users or break existing
+  automations, so be extremely careful here.
+
+  HPA creation/update with `minReplicas: 0` is no longer rejected.
```
This is only true if the enableScaleToZero field is also set to true, correct?

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,839 @@
+  1. Make sure there are no hpa objects with minReplicas=0 and maxReplicas=0. Here is a oneliner to update it to 1:
+
+      `$ kubectl get hpa --all-namespaces  --no-headers=true | awk  '{if($6==0) printf "kubectl patch hpa/%s --namespace=%s -p \"{\\\"spec\\\":{\\\"minReplicas\\\":1,\\\"maxReplicas\\\":1}}\"\n", $2, $1 }' | sh`
+  2. Disable `HPAScaleToZero` feature gate
```
On kube-apiserver and kubelet both, does it matter? Out of curiousity, why does kubelet need the feature gate?

What happens if you do this and there ARE HPA objects with min/max of 0/0?

johanneswuerbach replied:
> Out of curiousity, why does kubelet need the feature gate?

It seems not to be needed, removed.

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,839 @@
+
+  The e2e framework does not currently support enabling or disabling feature
+  gates. However, unit tests in each component dealing with managing data, created
+  with and without the feature, are necessary. At the very least, think about
```
Um, that's not an answer, it's just a copy of the comment above?

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,839 @@
+
+_This section must be completed when targeting beta graduation to a release._
+
+As this is a new field every usage is opt-in. In case the kubernetes version is downgraded, currently scaled to 0 workloads might not to be manually scaled to 1 as the controller would treat them as 
```
Here's where you would discuss what the operator needs to do to make sure there are no problems rolling this out. What happens when it's partially enabled (flag on different api servers / kubelets in different state, for example).

johnbelamaric reviewed (changes requested):
Looks like more work is needed on the PRR questions.

johanneswuerbach commented:
@johnbelamaric thanks for your review, updated the KEP.

MattJeanes commented:
/remove-lifecycle stale

noahlz commented:
Anything we the community can do to help move this along?

pbetkier commented:
@johnbelamaric could you have a look again as the author addressed your comments?

rotilho commented:
/assign @johnbelamaric

rotilho commented:
Shouldn't someone re-request @johnbelamaric? Since he's still in the `requested changes` status it won't appear in his review request queue.

ahg-g commented:
/cc

ofek commented:
Has there been any update on this?

johanneswuerbach commented:
/assign @johnbelamaric

Anything I could do here to move this forward?

noahlz commented:
We finally grew weary of waiting for this and migrated from HPA to KEDA. So far it's worked out pretty well.

Have already cut our AWS bill by ~30% (for the EKS cluster nodes), with more to come.

zroubalik commented:
> We finally grew weary of waiting for this and migrated from HPA to KEDA. So far it's worked out pretty well.
> 
> Have already cut our AWS bill by ~30% (for the EKS cluster nodes), with more to come.

Hi @noahlz! KEDA maintainer here. I would love to hear more about this, if you have a time, please. Could you please reach out (email on my GH account, Slack, LinkedIn, whatever) or tell me how can I connect with you? 
Thanks!

ofek commented:
/remove-lifecycle stale

MattJeanes commented:
/remove-lifecycle stale

pierluigilenoci commented:
/remove-lifecycle stale

MMeyer2808 commented:
Can anyone tell why this [pull request is not getting reviewed](https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#why-is-my-pull-request-not-getting-reviewed)?
Maybe it's just missing the `\lgtm` label as done before by e.g. @josephburnett, such that @johnbelamaric is notified to approve the requested changes?

> Once this PR has been reviewed and has the lgtm label, please ask for approval from johnbelamaric by writing /assign @johnbelamaric in a comment.

Would really love to see this feature flag reaching beta or GA :)

yongzhang commented:
Expecting this feature flag going to GA or at least beta for years!

adrianmoisey commented:
Hey @johanneswuerbach 
Are you still working on this? I'm keep to try help push it through, or to take it over from you if you're not willing.

adrianmoisey commented:
> @adrianmoisey not actively, I've tried to get this KEP reviewed, but there has been much activity. Happy to help with the implementation once there is movement here though.

Well... I'm not an approver, but I'm happy to help push this as much as I can.
Would it be possible to update the PR a bit, since it's quite out of date. 

Something I'm not 100% sure about, is that the feature seems to already exist, despite the KEP PR not being merged. But I'll do some digging and figure out the state of the world

adrianmoisey commented:
/ok-to-test

adrianmoisey commented on keps/sig-autoscaling/2021-scale-from-zero/kep.yaml:
```diff
@@ -0,0 +1,41 @@
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v1.16"
+  beta: "v1.27"
```
```suggestion
  beta: "v1.32"
```

adrianmoisey commented on keps/sig-autoscaling/2021-scale-from-zero/kep.yaml:
```diff
@@ -0,0 +1,41 @@
+# The most recent milestone for which work toward delivery of this KEP has been
+# done. This can be the current (upcoming) milestone, if it is being actively
+# worked on.
+latest-milestone: "v1.27"
```
```suggestion
latest-milestone: "v1.33"
```

adrianmoisey commented:
@johanneswuerbach can you make the suggested changes, then I'll ping some folks to try un-stuck this KEP

johanneswuerbach commented:
@adrianmoisey thanks, updated.

adrianmoisey commented:
@johanneswuerbach Sorry for the runaround. It's taking me a bit of time to get my head around where this KEP is in terms of process.

So... it's already implemented as alpha, and we need to promote it to beta.
There are a few sections in the KEP that are marked as required for beta, so, do you mind filling those in?

Again, sorry for the runaround!

And just to summarise what I think needs happening and the next steps:
1. Update KEP to beta
2. Get approval from sig-autoscaling chair
3. Get approval for PRR

johanneswuerbach commented:
@adrianmoisey I've updated the KEP template, are you talking about any specific section? Would be great reach out to sig-autoscaling again prior to iterating on this to understand what is exactly missing at this point.

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+extending the production code to implement this enhancement.
+-->
+
+- `<package>`: `<date>` - `<test coverage>`
```
```suggestion
- `/pkg/controller/podautoscaler`: `2025-02-06` - `96.4`
```

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- <test>: <link to test coverage>
```
N/A for our purposes?

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+- <test>: <link to test coverage>
```
We'll need to add e2e tests for this behaviour under https://github.com/kubernetes/kubernetes/tree/master/test/e2e/autoscaling, particularly given we're already mentioning this as a graduation to beta criteria below

johanneswuerbach replied:
Added as sentence, but not sure how to determine the test coverage here.

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+
+As this KEP changes the allowed values for `minReplicas`, special care is required for the downgrade case to not prevent any kind of updates for HPA objects using `minReplicas: 0`. The alpha code already accepts `minReplicas: 0` with the flag enabled or disabled since Kubernetes version 1.16 downgrades to any version >= 1.16 aren't an issue.
+
+The new flag `enableScaleToZero` defaults to `false`, which was has been the previous behavior. In flag should be disabled before downgrading as otherwise the
```
```suggestion
The new flag `enableScaleToZero` defaults to `false`, which was has been the previous behavior. The flag should be disabled before downgrading as otherwise the
```

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+automations, so be extremely careful here.
+-->
+
+  Any change of default behavior may be surprising to users or break existing
+  automations, so be extremely careful here.
```
```suggestion
```

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
```
```suggestion
```

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+This section must be completed when targeting beta to a release.
+-->
+
+As this is a new field every usage is opt-in. In case the kubernetes version is downgraded, currently scaled to 0 workloads might not to be manually scaled to 1 as the controller would treat them as
```
```suggestion
As this is a new field every usage is opt-in. In case the kubernetes version is downgraded, currently scaled to 0 workloads might need to be manually scaled to 1 as the controller would treat them as
```

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+are missing a bunch of machinery and tooling and can't do that now.
+-->
+
+No yet as no implementation is available.
```
Have we done this testing now that beta is being targeted?

johanneswuerbach replied:
If I understand correctly this points asks for the behavior of upgrading, enabling `enableScaleToZero` and downgrading. From what I understand this isn't really testable without an implementation 🤔

johnbelamaric replied:
don't you have an alpha implementation?

johanneswuerbach replied:
The alpha implementation works fairly different from the design proposed here and the idea behind this KEP was to address those shortcomings before promoting to beta.

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+logs or events for this purpose.
+-->
+
+The feature is used if workloads are scaled to zero by the autoscaling controller.
```
Can we provide a one liner for operators to examine the HPAs in their clusters for opt-in use of the `enableScaleToZero` field?

johnbelamaric replied:
If you have 50,000 clusters, a one liner is not really helpful. Metrics are best. Can kube state metrics help?

If it's not reasonable/practical to have a metric, then other means can be OK.

johanneswuerbach replied:
Yes, the property could be added to https://github.com/kubernetes/kube-state-metrics/blob/main/internal/store/horizontalpodautoscaler.go as an additional `kube_horizontalpodautoscaler_` metric.

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+Recall that end users cannot usually observe component logs or access metrics.
+-->
+
+Similar to autoscaling is confirmed today.
```
Can we provide any more detail here?

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+      - Impact of its degraded performance or high-error rates on the feature:
+-->
+
+The addition has the same dependencies as the currently autoscaling controller.
```
```suggestion
The addition has the same dependencies as the current autoscaling controller.
```

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+Are there any tests that were run/should be run to understand performance characteristics better
+and validate the declared limits?
+-->
+
```
```suggestion
No.
```

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
```
```suggestion
Autoscaling will not occur, this is the same as the current behaviour.
```

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
```
Not the failure mode (albeit the same as currently) where the controller is unable to fetch the relevant object or external metrics being used for scale from zero?

johanneswuerbach replied:
Done.

gjtempleton commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
```
As above.

johanneswuerbach replied:
Done.

gjtempleton reviewed (commented):
Mostly nits and typos, but some areas where we need to add some detail.

johanneswuerbach commented:
Thanks for the review @gjtempleton , I'll look into your questions until end of week.

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,934 @@
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `HPAScaleToZero`
+  - Components depending on the feature gate: `kube-apiserver`
```
kube-controller-manager? Where does HPA controller live?

johanneswuerbach replied:
Currently the gate `HPAScaleToZero` is only required on the `kube-apiserver`

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,934 @@
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `HPAScaleToZero`
+  - Components depending on the feature gate: `kube-apiserver`
+- [x] Other
```
i don't think you need to fill this section, the feature gate is sufficient

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,934 @@
+-->
+
+
+Yes. To downgrade the cluster to version that does not support scale-to-zero feature:
```
or if you turn off the feature gate?

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,934 @@
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Nothing, the feature can be re-enabled without problems.
```
for example, will "stuck at 0" workloads start getting scaled again?

johanneswuerbach replied:
Added.

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,934 @@
+will rollout across nodes.
+-->
+
+There are no expected side-effects when the rollout fails as new `enableScaleToZero` flag should only be enabled once the version upgraded completed and should be disabled before attempting a rollback.
```
If the new code in the hpa controller crashes, will it block other HPAs that don't use the feature?

johanneswuerbach replied:
I'm not entirely certain about the details of panic handling in k8s controllers, but the HPA controller implements essentially a queue with multiple workers, so if one of those workers consistently crashes, other HPAs would be affected as well unless the default k8s queue implementation somehow protects against this.

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,934 @@
+that might indicate a serious problem?
+-->
+
+If workloads aren't scaled up from 0 despite the scaling condition being meet, an operator should rollback this feature and manually scale an affected workload back to `1`.
```
Is there a metric or something? Imagine you run 50,000 clusters and your users may or may not be using this feature you enabled. How can you easily tell if it's working or needs to be turned off? Typically this is a metric.

johanneswuerbach replied:
Working autoscaling can generally be verified by comparing [kube_horizontalpodautoscaler_status_current_replicas](https://github.com/kubernetes/kube-state-metrics/blob/main/internal/store/horizontalpodautoscaler.go#L303C4-L303C56) and [kube_horizontalpodautoscaler_status_desired_replicas](https://github.com/kubernetes/kube-state-metrics/blob/main/internal/store/horizontalpodautoscaler.go#L391C4-L391C49) and some alerting when those values aren't matching within a timeframe. Scale to zero could be observed in a similar fashion.

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,934 @@
+Even if applying deprecation policies, they may still surprise some users.
+-->
+
+Once this is in beta, the alpha flag can be removed.
```
you can just say "No" here

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,934 @@
+implementation difficulties, etc.).
+-->
+
+No, in regards to this KEP.
```
see comments above...

Would it be useful to track how often scaling to/from zero happens?

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,934 @@
+  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
+-->
+
+Yes, one additional boolean field.
```
to what resource?

johanneswuerbach commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+## Summary
+
+[Horizontal Pod Autoscaler][] (HPA) automatically scales the number of pods in any resource which supports the `scale` subresource based on observed CPU utilization
+(or, with custom metrics support, on some other application-provided metrics) from one to many replicas. This proposal adds support for scaling from zero to many replicas and back to zero for object and external metrics.
```
Added, better?

johanneswuerbach commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,939 @@
+provided metric. A typical use-case case for this is scaling the number of queue consumers based on the length of the consumed queue.
+
+In cases of a frequently idle queue or a less latency sensitive workload there is no need to run one replica at all times and instead you want to dynamically scale
+to zero replicas, especially if those replicas have high resource requests. If replicas are scale to 0, HPA also needs the ability to scale up once messages are available.
```
Added as well.

johanneswuerbach commented:
Thanks for the feedback, pushed some updates.

adrianmoisey commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,955 @@
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+Currently disabling HPA is possible by manually setting the scaled resource to `replicas: 0`. This works as the HPA itself could never reach this state itself.
+As `replicas: 0` is now a possible state when using `minReplicas: 0` it can no longer be used to differentiate between manually disabled or automatically scaled to zero.
+
+Additionally the `replicas: 0` state is problematic as updating a HPA object `minReplicas` from `0` to `1` has different behavior. If `replicas` was `0` during the update, HPA
+will be disabled for the resource, if it was `> 0`, HPA will continue with the new `minReplicas` value.
+
+To resolve this issue the KEP is introducing an explicit `enableScaleToZero` property to explicitly enable/disable scale from/to zero.
```
What if instead of having a `enableScaleToZero` property, we allow HPAs to scale to 0 if they are for objects or external metrics, and introduce a `disabled` flag, allowing someone to disable the HPA completely?

In the world with `enableScaleToZero` enabled, how would someone disable that HPA?

tcolgate replied:
> What if instead of having a `enableScaleToZero` property, we allow HPAs to scale to 0 if they are for objects or external metrics, and introduce a `disabled` flag, allowing someone to disable the HPA completely?
> 
> In the world with `enableScaleToZero` enabled, how would someone disable that HPA?

Is it even try that 0 disabled? Last time I tested, 0 was rejected as invalid, there was no option for disabling (I had to delete HPAs to "temporarily disabled auto scaling"

(I'm not in a position to test this right now).

FWIW, I think an explicit "allow scaling to zero" flag is a better option anyway. And an explicit disable flag would be very useful regardless of scale to zero

adrianmoisey replied:
> Is it even true that 0 disables? Last time I tested, 0 was rejected as invalid, there was no option for disabling (I had to delete HPAs to "temporarily disabled auto scaling"

On which resource specifically?

The disabling of the HPA is when the scaled resource's replicas is set to zero, see: 
- https://github.com/kubernetes/kubernetes/blob/1d73c5d882fdb70037e5944cb9492415b4187774/pkg/controller/podautoscaler/horizontal.go#L817-L821
- https://github.com/kubernetes/kubernetes/pull/29212/files


> FWIW, I think an explicit "allow scaling to zero" flag is a better option anyway. 

Why so? Isn't setting `minReplicas: 0` an explicit way to say "I'm Ok with scaling down to zero pods".
We can also safe-guard it by only allowing objects or external metrics.

omerap12 replied:
Like Adrian said, setting `minReplicas: 0` is an explicit way to tell the system it's okay to scale down to zero pods. 
IMHO, that shouldn’t have anything to do with turning the HPA on or off.

johanneswuerbach replied:
As the KEP states the problem is the changing from `minReplicas: 0` to `minReplicas: 1` (or any value > 0). In this case the HPA won't perform the change from `replicas: 0` ~> `replicas: 1` as currently `minReplicas: 1` and `replicas: 0` is a way to signal "stop autoscaling".

This could be change, but is technically a breaking change hence the added property to explicitly override this behaviour.

adrianmoisey commented on keps/sig-autoscaling/2021-scale-from-zero/kep.yaml:
```diff
@@ -0,0 +1,42 @@
+replaces:
+
+# The target maturity stage in the current dev cycle for this KEP.
+stage: beta
```
I think this may need to stay as alpha, since `enableScaleToZero` is being introduced.
See: https://github.com/kubernetes/community/blob/ecfd8bbf586d7bcd9767171c4055f2cf6502bb85/contributors/devel/sig-architecture/feature-gates.md?plain=1#L80-L81

And:
https://github.com/kubernetes/community/blob/ecfd8bbf586d7bcd9767171c4055f2cf6502bb85/contributors/devel/sig-architecture/feature-gates.md?plain=1#L137-L150

The scenario I'm worried about is that someone uses `enableScaleToZero` in 1.33 (I know the feature missed 1.33, this is just an example) when the feature gate is set to beta, and enabled by default.

They then downgrade to 1.32, and now the API server fails on HPA resources with `enableScaleToZero` set. 


If `enableScaleToZero` had been added in 1.32 with the alpha feature gate defaulted to off, this wouldn't have been a problem.

johanneswuerbach replied:
Sounds good, updated.

ofek commented:
/remove-lifecycle stale

jm-franc commented:
To recap:

1. There was a problem related to updating from `minReplicas:0` to `:1` (see [here](https://github.com/kubernetes/enhancements/pull/2022#discussion_r2148811309)), but Jordan gave a nice solution to that (i.e. adding a `scaledToZero` condition, see [here](https://github.com/kubernetes/enhancements/pull/2022#discussion_r2180962074)).
2. @johnbelamaric asked for a few clarifications (e.g. [here](https://github.com/kubernetes/enhancements/pull/2022#discussion_r1955178339)), but nothing problematic as far as I can see.
3. It looks like there's a consensus about adding a 'paused' field to allow users to pause the HPA, to be added in a follow-up KEP.

It looks like we're getting there :) @johanneswuerbach I can help you with those updates if you want me to.

johanneswuerbach commented:
@jm-franc thank you. I've incorporated the changes around the status and addressed some open questions, but can't resolved them in the GitHub UI . 

Let me know if I should update something else.

johanneswuerbach commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,952 @@
+const (
+ // ScaledToZero indicates that the HPA controller scaled the workload to zero.
+ ScaledToZero HorizontalPodAutoscalerConditionType = "ScaledToZero"
+)
```
Added.

adrianmoisey commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,952 @@
+nitty-gritty.
+-->
+
+Allow the HPA to scale from and to zero using `minReplicas: 0` when explicitly enabled with a flag.
```
Nit: what is "with a flag"? Is that left over from the previous iteration before the condition was used to mark the scaled-to-zero? Or is that the feature gate?

jm-franc commented:
> @jm-franc thank you. I've incorporated the changes around the status and addressed some open questions, but can't resolved them in the GitHub UI .
> 
> Let me know if I should update something else.

To me this looks ready to go, thank you @johanneswuerbach ! As Adrian mentioned, it would be nice to have somebody from @kubernetes/sig-autoscaling-leads review this so this can be included in 1.35. I'll mention it in the next SIG autoscaling meeting.

/lgtm

adrianmoisey commented:
Hey  @johnbelamaric 

Can you re-review this from a PRR perspective?

johanneswuerbach commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,952 @@
+Additionally the `replicas: 0` state is problematic as updating a HPA object `minReplicas` from `0` to `1` has different behavior. If `replicas` was `0` during the update, HPA
+will be disabled for the resource, if it was `> 0`, HPA will continue with the new `minReplicas` value.
+
+To resolve these issues the KEP is introducing an explicit `ScaledToZero` condition inside the `HorizontalPodAutoscalerStatus`. When `ScaledToZero=True` was recorded the HPA will scale
```
The problem we also saw with the current alpha implementation is that you can't modify an HPA from `minReplicas: 0 ~> 1` if the scaling target is currently at `replicas: 0` as `minReplicas: 1 & replicas: 0` means HPA paused and the targeted won't be scaled anymore, which is not obvious.

The condition is primarily added to differentiate between this explicit pause state and an implicitly reached pause state, so `minReplicas: 0 ~> 1` behaves exactly like `minReplicas: 1 ~> 2` and scales the target in both cases.

jackfrancis replied:
> With your idea users would need to ensure minReplicas > 0

This makes no sense in the context of today because `minReplicas: 0` is undefined. There is no back-compat story for any version of configuration that includes `minReplicas: 0`.

omerap12 replied:
AFAIK this behavior isn't officially documented but it's a widely used pattern in practice. And like my friend @adrianmoisey like to send: https://www.hyrumslaw.com/

liggitt replied:
> At present, `minReplicas: 0` is not a valid HPA configuration, so simply enabling that as a valid configuration should solve the differentiation problem.

Because HPA is mutable, an existing HPA can change from `minReplicas: 0` → `minReplicas: 1`. When it does, without the condition, we cannot tell if the `replicas: 0` on the object means the user disabled it and HPA should ignore it, or HPA was responsible for scaling it to 0 and should scale it up to meet the new `minReplicas`.

liggitt replied:
> There is no official documentation saying this, but, it was intentionally added

Explicitly added in https://github.com/kubernetes/kubernetes/issues/28603, and definitely relied on at this point.

jackfrancis replied:
> I'm not convinced we need a new ScaledToZero condition.

I've been convinced. 🙂

johnbelamaric reviewed (commented):
PRR

Last commit was in July? Has this been updated with the latest PRR template? Are you still targeting 1.35?

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -143,138 +181,142 @@ updates.
+-->
 
-- Allow users to optionally override the default workload autoscaling tolerance on a per-HPA bases.
+Allow the HPA to scale from and to zero using `minReplicas: 0` and a HPA status condition.
```
a little more detail would be helpful

johanneswuerbach replied:
Could you elaborate whether you mean more detail why this is needed or how this is implemented?

The mechanism is currently described below, but I can move it here as well.

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -490,8 +556,22 @@ well as the [existing list] of feature gates.
-  - Feature gate name: HPAConfigurableTolerance
-  - Components depending on the feature gate: `kube-controller-manager`
+  - Feature gate name: `HPAScaleToZero`
+  - Components depending on the feature gate: `kube-apiserver`
```
where does the hpa controller live? kcm?

johanneswuerbach replied:
The HPA controller runs inside the KCM. The flag is only needed in the API server, as the HPA already supports scaling down to zero, but only `minReplicas > 0` is allowed by the API server without this flag. The flag itself is already implemented since 1.16

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -530,20 +614,19 @@ The e2e framework does not currently support enabling or disabling feature
 -->
 
+As this is a new field every usage is opt-in. In case the kubernetes version is downgraded, currently scaled to 0 workloads might need to be manually scaled to 1 as the controller would treat them as
+paused otherwise.
```
recommendation on ordering of enablement/rollout? what happens if someone enables it in api server, someone uses it, but hpa has not been upgraded yet?

johanneswuerbach replied:
This is mentioned above:


> 1. Make sure there are no hpa objects with minReplicas=0 and maxReplicas=0. Here is a oneliner to update it to 1:
> 
>     `$ kubectl get hpa --all-namespaces  --no-headers=true | awk  '{if($6==0) printf "kubectl patch hpa/%s --namespace=%s -p \"{\\\"spec\\\":{\\\"minReplicas\\\":1,\\\"maxReplicas\\\":1}}\"\n", $2, $1 }' | sh`
> 2. Disable `HPAScaleToZero` feature gate
> 3. In case step 1. has been omitted, workloads might be stuck with `replicas: 0` and need to be manually scaled up to `replicas: 1` to re-enable autoscaling.

I can reference it here again.

johnbelamaric commented on keps/sig-autoscaling/2021-scale-from-zero/README.md:
```diff
@@ -0,0 +1,839 @@
+###### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
```
unresolved, I don't think the response really answers the question

johanneswuerbach commented on keps/sig-autoscaling/2021-scale-from-zero/kep.yaml:
```diff
@@ -22,23 +21,22 @@ stage: alpha
-  alpha: TBD
-  beta: TBD
-  stable: TBD
+  alpha: "v1.16"
```
Yes, this was already released as alpha in 1.16. At this time no KEP was written and this KEP is about documenting this and updating the existing implementation based on learnings from the alpha.

Once those changes are implemented, a beta release can be attempted.

johnbelamaric commented:
ok, I can approve the PRR once sig autoscaling gives their approval - see https://github.com/kubernetes/enhancements/pull/2022#issuecomment-2596815237

johnbelamaric commented:
/approve
/lgtm
